### PR TITLE
Updates for infallible creation methods

### PR DESCRIPTION
**Pull request by danieldg**
This is the SCTK side of https://github.com/Smithay/wayland-rs/pull/531 and https://github.com/Smithay/wayland-rs/pull/530 - it moves failures out of the `Arc::new_cyclic` closure, removing the need for the `Option` there.

### CODE REVIEW

File: src/compositor.rs [hunk 1]
```diff
@@ -85,7 +85,7 @@ impl CompositorState {
     {
         let compositor = self.wl_compositor.get()?;
 
-        let surface = compositor.create_surface(qh, data)?;
+        let surface = compositor.create_surface(qh, data);
```
Comment (i509VCB) [inline src/compositor.rs:88]: Does GlobalError still need an `InvalidId` variant?

**Review by i509VCB — CHANGES_REQUESTED** [review]

File: src/compositor.rs (hunk 1 above)
Comment (danieldg) [reply src/compositor.rs:88]: We can probably remove all of the `InvalidId` variants after this patch, and maybe also some other errors (shm create_buffer can be infallible now).  That might be useful to add in this series.

File: src/shell/layer/mod.rs [hunk 2]
```diff
@@ -127,38 +127,24 @@ impl LayerSurfaceBuilder {
     {
         // The layer is required in ext-layer-shell-v1 but is not part of the factory request. So the param
         // will stay for ext-layer-shell-v1 support.
+        let layer_shell = shell.bound_global()?;
+        let freeze = qh.freeze();
```
Comment (i509VCB) [inline src/shell/layer/mod.rs:131]: Let's explain why we need to freeze the queue during construction.

File: src/shell/xdg/popup.rs [hunk 3]
```diff
@@ -80,46 +78,45 @@ impl Popup {
...
         });
-        // This assert checks that err was set properly above; it should be impossible to trigger,
-        // so it's not a run-time assert.
-        debug_assert!(inner.is_some() || err.is_err());
-        err.map(|()| Popup { inner })
+        drop(freeze);
+        Ok(Popup { inner })
     }
 
     fn inner(&self) -> &PopupInner {
```
Comment (i509VCB) [inline src/shell/xdg/popup.rs:118]: This function is effectively redundant now.

**Review by i509VCB — APPROVED** [review]
This looks good, need to get CI to be green.

File: src/shell/xdg/popup.rs (hunk 3 above)
Comment (danieldg) [reply src/shell/xdg/popup.rs:118]: Which function? `from_surface` is not redundant with `new`.

Comment (i509VCB) [reply src/shell/xdg/popup.rs:120]: The `inner` functions

Comment (danieldg) [reply src/shell/xdg/popup.rs:120]: Ah. The github quote function makes this very unclear: it shows lines 117-121, and your comment is about line 121-123.

**Pull request merged**